src/builder/build_traits.rs

```diff
@@ -62,14 +62,25 @@ pub trait Container {
         Ok(Wire::new(src, src_port))
     }
 
-    /// Add a constant value to the container and return a handle to it.
+    /// Add a constant value to the container and return a handle to it. The
+    /// `provenance` argument says which extension the type and value of the
+    /// constant comes from. If it is `None`, then the constant must be a built
+    /// in type.
     ///
     /// # Errors
     ///
     /// This function will return an error if there is an error in adding the
     /// [`OpType::Const`] node.
-    fn add_constant(&mut self, constant: ops::Const) -> Result<ConstID, BuildError> {
-        let const_n = self.add_child_op(constant)?;
+    fn add_constant(
+        &mut self,
+        constant: ops::Const,
+        provenance: Option<ExtensionId>,
```

Comment:
Didn't we say 'provenance' should not be an Id but a Set? I'm not sure whether an Option<ExtensionSet> or just an ExtensionSet, depending on whether None means "no extensions" (=empty ExtensionSet) or "unknown extensions" (different from an empty ExtensionSet)...

Reply:
Oh, again this might be the wrong PR